File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/displayable/HeaderRowDisplayable.java
package com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable;

import com.benoitquenaudon.tvfoot.red.R;
import com.benoitquenaudon.tvfoot.red.util.DateUtilsKt;
import com.google.auto.value.AutoValue;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import kotlin.text.StringsKt;
import timber.log.Timber;

@AutoValue public abstract class HeaderRowDisplayable implements MatchesItemDisplayable {
  private static final ThreadLocal<DateFormat> headerKeyDateFormat = new ThreadLocal<DateFormat>() {
    @Override protected DateFormat initialValue() {
      SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
      format.setTimeZone(TimeZone.getDefault());
      return format;
    }
  };
  private static final ThreadLocal<DateFormat> monthDateFormat = new ThreadLocal<DateFormat>() {
    @Override protected DateFormat initialValue() {
      SimpleDateFormat format = new SimpleDateFormat("EEEE, d MMMM", Locale.getDefault());
      format.setTimeZone(TimeZone.getDefault());
      return format;
    }
  };
  private static final ThreadLocal<DateFormat> yearDateFormat = new ThreadLocal<DateFormat>() {
    @Override protected DateFormat initialValue() {
      SimpleDateFormat format = new SimpleDateFormat("EEEE, d MMMM yyyy", Locale.getDefault());
      format.setTimeZone(TimeZone.getDefault());
      return format;
    }
  };

  public abstract int dangerResId();

  public abstract boolean hasDanger();

  public abstract String displayedDate();

  public static HeaderRowDisplayable create(String headerKey) {
    Date date;
    try {
      date = headerKeyDateFormat.get().parse(headerKey);
    } catch (ParseException e) {
      Timber.e(e);
      throw new UnsupportedOperationException("What is this date anyway? " + headerKey);
    }

    int dangerResId = -1;
    String displayedDate;
    Calendar nowCalendar = Calendar.getInstance();
    if (DateUtilsKt.isToday(date.getTime(), nowCalendar)) {
      dangerResId = R.string.matches_row_header_danger_today;
      displayedDate = StringsKt.capitalize(monthDateFormat.get().format(date));
      return new AutoValue_HeaderRowDisplayable(dangerResId, true, displayedDate);
    }

    if (DateUtilsKt.isTomorrow(date.getTime(), nowCalendar)) {
      dangerResId = R.string.matches_row_header_danger_tomorrow;
      displayedDate = StringsKt.capitalize(monthDateFormat.get().format(date));
      return new AutoValue_HeaderRowDisplayable(dangerResId, true, displayedDate);
    }

    if (DateUtilsKt.isCurrentYear(date.getTime(), nowCalendar)) {
      displayedDate = StringsKt.capitalize(monthDateFormat.get().format(date));
    } else {
      displayedDate = yearDateFormat.get().format(date);
    }

    return new AutoValue_HeaderRowDisplayable(dangerResId, false, displayedDate);
  }

  @Override public boolean isSameAs(MatchesItemDisplayable newItem) {
    return newItem instanceof HeaderRowDisplayable && this.displayedDate()
        .equals(((HeaderRowDisplayable) newItem).displayedDate());
  }
}

File: app/src/main/java/com/benoitquenaudon/tvfoot/red/app/domain/matches/displayable/HeaderRowDisplayable.kt
package com.benoitquenaudon.tvfoot.red.app.domain.matches.displayable

import com.benoitquenaudon.tvfoot.red.R
import com.benoitquenaudon.tvfoot.red.util.formatter
import com.benoitquenaudon.tvfoot.red.util.isCurrentYear
import com.benoitquenaudon.tvfoot.red.util.isToday
import com.benoitquenaudon.tvfoot.red.util.isTomorrow
import timber.log.Timber
import java.text.DateFormat
import java.text.ParseException
import java.util.Calendar
import java.util.Date

@Suppress("DataClassPrivateConstructor")
data class HeaderRowDisplayable private constructor(
    /**
     * Used for sticky headers
     */
    val id: Int,
    val dangerResId: Int,
    val hasDanger: Boolean,
    val displayedDate: String
) : MatchesItemDisplayable {
  override fun isSameAs(other: MatchesItemDisplayable): Boolean {
    return other is HeaderRowDisplayable && this.displayedDate == other.displayedDate
  }

  companion object Factory {
    private val headerKeyDateFormat: DateFormat by formatter("yyyy-MM-dd")
    private val monthDateFormat: DateFormat by formatter("EEEE, d MMMM")
    private val yearDateFormat: DateFormat by formatter("EEEE, d MMMM yyyy")

    fun create(headerKey: String): HeaderRowDisplayable {
      val date: Date
      try {
        date = headerKeyDateFormat.parse(headerKey)
      } catch (e: ParseException) {
        Timber.e(e)
        throw UnsupportedOperationException("What is this date anyway? " + headerKey)
      }

      // yyyy-MM-dd without the '-' to int
      val id = headerKeyDateFormat.format(date).filter { it != '-' }.toInt()

      var dangerResId = -1
      val displayedDate: String
      val nowCalendar = Calendar.getInstance()
      if (date.time.isToday(nowCalendar)) {
        dangerResId = R.string.matches_row_header_danger_today
        displayedDate = monthDateFormat.format(date).capitalize()
        return HeaderRowDisplayable(id, dangerResId, true, displayedDate)
      }

      if (date.time.isTomorrow(nowCalendar)) {
        dangerResId = R.string.matches_row_header_danger_tomorrow
        displayedDate = monthDateFormat.format(date).capitalize()
        return HeaderRowDisplayable(id, dangerResId, true, displayedDate)
      }

      displayedDate = if (date.time.isCurrentYear(nowCalendar)) {
        monthDateFormat.format(date).capitalize()
      } else {
        yearDateFormat.format(date)
      }

      return HeaderRowDisplayable(id, dangerResId, false, displayedDate)
    }
  }
}
